packages/ai-bot/main.ts

```diff
@@ -162,56 +222,106 @@ async function sendStream(
-      messages: [{ role: 'user', content: content.body }],
-      stream: true,
+function constructHistory(history: IRoomEvent[]) {
+  const events = new Map<string, IRoomEvent[]>();
+  for (let event of history) {
+    let content = event.content;
+    if (event.type == 'm.room.message') {
+      let event_id = event.event_id!;
+      if (content['m.relates_to']?.rel_type === 'm.replace') {
+        event_id = content['m.relates_to']!.event_id!;
+      }
+      if (!events.get(event_id)) {
+        events.set(event_id, [event]);
+      } else {
+        events.get(event_id)!.push(event);
+      }
+    }
+  }
+  let latest_events: IRoomEvent[] = [];
+  events.forEach((event_list, _event_id) => {
+    event_list = event_list.sort((a, b) => {
+      return a.origin_server_ts - b.origin_server_ts;
     });
+    latest_events.push(event_list[event_list.length - 1]);
+  });
+  latest_events = latest_events.sort((a, b) => {
+    return a.origin_server_ts - b.origin_server_ts;
+  });
+  return latest_events;
+}
+
+function getLastUploadedCardID(history: IRoomEvent[]): String | undefined {
+  for (let event of history.slice().reverse()) {
+    const content = event.content;
+    if (content.msgtype === 'org.boxel.card') {
+      let card = content.instance.data;
+      return card.id;
+    }
   }
+  return undefined;
+}
+
+async function getResponse(history: IRoomEvent[]) {
+  let historical_messages: Message[] = [];
+  log.info(history);
+  for (let event of history) {
+    let body = event.content.body;
+    log.info(event.sender, body);
+    if (body) {
+      if (event.sender === 'aibot') {
+        historical_messages.push({
+          role: 'assistant',
+          content: body,
+        });
+      } else {
+        historical_messages.push({
+          role: 'user',
+          content: getUserMessage(event),
+        });
+      }
+    }
+  }
+  let messages: Message[] = [
+    {
+      role: 'system',
+      content: MODIFY_SYSTEM_MESSAGE,
+    },
+  ];
+
+  messages = messages.concat(historical_messages);
+  log.info(messages);
+  return await openai.chat.completions.create({
+    model: 'gpt-4-0613',
+    messages: messages,
+    stream: true,
+  });
 }
 
 (async () => {
-  let client = createClient({ baseUrl: 'http://localhost:8008' });
-  let auth = await client.loginWithPassword('aibot', 'pass');
+  let client = createClient({
+    baseUrl: process.env.MATRIX_URL || 'http://localhost:8008',
+  });
+  let auth = await client.loginWithPassword(
+    'aibot',
+    process.env.BOXEL_AIBOT_PASSWORD || 'pass'
+  );
   let { user_id } = auth;
   client.on(RoomMemberEvent.Membership, function (_event, member) {
     if (member.membership === 'invite' && member.userId === user_id) {
       client.joinRoom(member.roomId).then(function () {
-        console.log('Auto-joined %s', member.roomId);
+        log.info('Auto-joined %s', member.roomId);
       });
     }
   });
+
   // TODO: Set this up to use a queue that gets drained
```

Comment:
you'll definitely still want to handle your promises here--it will be very difficult to run in tests without keeping track of promises

Reply:
@habdelra does this comment have to do with the fact that matrix event handler wud not necessarily support an async function 

ie some change like this

```

  client.on(RoomEvent.Timeline, function (event, room, toStartOfTimeline) { //function no longer async
    handleRoomEvent(client, event, room, toStartOfTimeline); //handleRoomEvent is not async and calls a IFFE
  });
```

Reply:
even if we call an IFFE we are still not keeping track of the promise. The idea with using an event queue that gets drained is that we track all the promises. This way if we make nodejs tests against the AI bot we'll know when the server state has settled before starting a new test. otherwise a running promise from a previous test could bleed into a subsequent test

Reply:
An example of how we should handle this is the matrix handlers in packages/host/lib/matrix-handlers/timeline.ts and membership.ts. In both of these we want to perform async work as part of a synchronous subscription to matrix events. For example in membership.ts, there is a synchronous `onMembership()` event callback that we wire into the matrix on membership event. The idea is that we then take each event and push it on to a queue of pending work. We then synchonously call a function that kicks off the draining of the queue (in this case we also debounce that drain too for more efficient performance). In the drain we await any previous work and then we save a stateful promise for the current work, which we fulfill (in a deferred like fashion) at the end of the drain. That stateful promise that we save is the promise that we were otherwise throwing away (the leaky async). We now have a way to "see" it and await it in testing.